Feature/add rqtl endpoints

zsloan opened the pull request:
#### Description
This PR will add endpoint(s) for calling the R script that will run R/qtl (in place of using rpy2). Modeling it off of how Bonface did the GEMMA endpoints, except I'm using POST-ed form data due to the much larger number of arguments.

#### How should this be tested?
Requests should be made to the relevant endpoints; for example:
curl -F "geno_file=tests/test_data/BXD.geno" -F "pheno_file=tests/test_data/BXD_pheno.csv" http://localhost:8080/api/rqtl/compute

#### Any background context you want to provide?

#### What are the relevant pivotal tracker stories?

#### Screenshots (if appropriate)

#### Questions

BonfaceKilz committed:
db: phenotypes: Add a way of updating the Phenotype table

* gn3/db/phenotypes.py (Phenotype): New dataclass.
(update_phenotype): New function.
[phenotype_column_mapping]: New variable.

BonfaceKilz committed:
tests: test_phenotypes: New test cases for loading phenotypes

zsloan committed:
Register rqtl blueprint

zsloan committed:
Add rqtl.py for rqtl endpoints

zsloan committed:
Added RQTL_WRAPPER_CMD (which is fetched from environment) in settings.py

zsloan committed:
Read in kwargs fromrequest and pass the command, tmpdir, and kwargs to generate_rqtl_cmd which returns the actual command and output path

zsloan committed:
Created compose_rqtl_command and generate_rqtl_command to create the actual command to be run from the command line; used the same pattern as for GEMMA for consistency

robwwilliams commented:
Good luck Zach!

On Mon, May 17, 2021 at 5:03 PM zsloan ***@***.***> wrote:

> Description
>
> This PR will add endpoint(s) for calling the R script that will run R/qtl
> (in place of using rpy2). Modeling it off of how Bonface did the GEMMA
> endpoints, except I'm using POST-ed form data due to the much larger number
> of arguments.
> How should this be tested?
>
> Requests should be made to the relevant endpoints; for example:
> curl -F "geno_file=tests/test_data/BXD.geno" -F
> "pheno_file=tests/test_data/BXD_pheno.csv"
> http://localhost:8086/api/rqtl/compute
> Any background context you want to provide? What are the relevant pivotal
> tracker stories? Screenshots (if appropriate) Questions
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/genenetwork/genenetwork3/pull/13
> Commit Summary
>
>    - Register rqtl blueprint
>    - Add rqtl.py for rqtl endpoints
>    - Added RQTL_WRAPPER_CMD (which is fetched from environment) in
>    settings.py
>    - Read in kwargs fromrequest and pass the command, tmpdir, and kwargs
>    to generate_rqtl_cmd which returns the actual command and output path
>    - Created compose_rqtl_command and generate_rqtl_command to create the
>    actual command to be run from the command line; used the same pattern as
>    for GEMMA for consistency
>
> File Changes
>
>    - *A* gn3/api/rqtl.py
>    <https://github.com/genenetwork/genenetwork3/pull/13/files#diff-b664cc0807e00a544ba0102495aa7feac8c95cc4f4499f1765025e18aa2d7c46>
>    (37)
>    - *M* gn3/app.py
>    <https://github.com/genenetwork/genenetwork3/pull/13/files#diff-7f0d78685ebf612a64b1eabea3410db650ccc34c707603c682673d5e541e85fb>
>    (2)
>    - *M* gn3/commands.py
>    <https://github.com/genenetwork/genenetwork3/pull/13/files#diff-c0443760c217813cb8614b1629640ff4919939a39199e8625dc2aed7a2ffb874>
>    (7)
>    - *A* gn3/computations/rqtl.py
>    <https://github.com/genenetwork/genenetwork3/pull/13/files#diff-2f07fc1fb8bcb779abe3d5fb06736cb465dadc2c2205783bd117a4f726776cc9>
>    (25)
>    - *M* gn3/settings.py
>    <https://github.com/genenetwork/genenetwork3/pull/13/files#diff-209b9b2950a5be2be9d1ed13b72f8a82fe4e9b6c6c2e9fcd6f3d22b550703cce>
>    (1)
>
> Patch Links:
>
>    - https://github.com/genenetwork/genenetwork3/pull/13.patch
>    - https://github.com/genenetwork/genenetwork3/pull/13.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork3/pull/13>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7GMFAPPIOTIRAGKZADTOGHE5ANCNFSM45BJKF3A>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: ***@***.***  or  ***@***.***

zsloan committed:
Fixed variety of issues detected by pylint

BonfaceKilz reviewed (approved):
LGTM! Usually it's also a nice idea to make it a habit to write tests for new functionality you add \m/\m/

zsloan commented:
> LGTM! Usually it's also a nice idea to make it a habit to write tests for new functionality you add \m/\m/

Thanks for reminding me; I'll try to add some tests today.

zsloan committed:
Fixed generate_rqtl_cmd to make the kwarg hash from a combination of keywords and arguments + account for boolean kwargs without values (like --interval or --covar)

zsloan committed:
Account for boolean kwargs in compose_rqtl_cmd

zsloan committed:
Account for boolean kwargs in compute() by storing them in a list, since they don't have corresponding values

zsloan committed:
Added unit test for computations/rqtl.py

zsloan committed:
Removed len from this if statement, since an empty list evaluates to False by itself

zsloan committed:
Added test for compose_rqtl_cmd in tests/unit/test_commands.py

BonfaceKilz committed:
db: phenotype: Make "pylint: disable=[R0902]" global for file

BonfaceKilz committed:
db: phenotypes: Add dataclass to represent PublishXRef

BonfaceKilz committed:
db: phenotypes: Add phenotype table mapping

BonfaceKilz committed:
db: phenotypes: Put mapping def after dataclass

zsloan committed:
Merge branch 'main' of https://github.com/genenetwork/genenetwork3 into feature/add_rqtl_endpoints

BonfaceKilz committed:
db: phenotypes: Add type for Dataclass

See: https://www.py4u.net/discuss/188952

BonfaceKilz committed:
db: phenotypes: Map a table to it's relevant dict mapping

BonfaceKilz committed:
db: phenotypes: Generalise the update function

* gn3/db/phenotypes.py (update_phenotype): Delete it.
(update): New, more general function.

BonfaceKilz committed:
db: phenotypes: Update failing tests

BonfaceKilz committed:
db: phenotypes: Add Publication dataclass and mapping

zsloan committed:
Temporarily replace forward-slashes with underscores, since they can be included in the hashes used for filenames

zsloan committed:
Merge branch 'main' of https://github.com/genenetwork/genenetwork3 into feature/add_rqtl_endpoints

BonfaceKilz committed:
db: phenotypes: Rename phenotype_column_mapping

BonfaceKilz committed:
db: phenotypes: Add Publication table mapping

BonfaceKilz committed:
db: phenotypes: Add a dataclass map

Maps a string to it's dataclass.

BonfaceKilz committed:
db: phenotypes: Add function for fetching a single result

* gn3/db/phenotypes.py (fetchone): New function.

BonfaceKilz committed:
tests: test_phenotype: Add function that tests "fetchone"

BonfaceKilz committed:
db: phenotypes: Fix typo

BonfaceKilz committed:
db: phenotypes: Fix publication_mapping

zsloan committed:
Include code that processes rqtl output files and returns actual results instead of just the output filename

zsloan committed:
Fix R/qtl command and the way keyword arguments are passed

zsloan committed:
Add rqtl_wrapper.R script and necessary imports to guix.scm

BonfaceKilz committed:
Move the methods, "update" and "fetch", to gn3.db

zsloan committed:
Added option for enabling permutation strata

Fixed issue where covariates with numerical names were being ignored

Fixed issue where the hash filename wasn't being used for the permutation output

zsloan committed:
Fixed issue where all bool kwargs were always being passed to generate_rqtl_cmd and also made code check if output file already exists (so caching works)

Alexanderlacuna committed:
fix index error (#16)

BonfaceKilz committed:
gn3: db: Return None if data and where are empty

BonfaceKilz committed:
Get the diff between 2 dicts and return that as a dict

BonfaceKilz committed:
unit: test_phenotypes: Test for multiple "WHERE" clauses

BonfaceKilz committed:
gn3: db: Add spacing before around "AND" in sql clause

BonfaceKilz committed:
Add data structures for the table metadata_audit

BonfaceKilz committed:
sql: Add a new folder that houses sql statements for new tables

BonfaceKilz committed:
gn3: db: Add new function for doing sql INSERT

BonfaceKilz committed:
tests: test_audit: Update test case to check for prepared statements

BonfaceKilz committed:
Use prepared statements for UPDATE sql function

BonfaceKilz committed:
gn3: db: Replace items() with keys()

* gn3/db/__init__.py (diff_from_dict): We only use the keys of the dict!

BonfaceKilz committed:
Use prepared statements for FETCH sql function

BonfaceKilz committed:
gn3: db: Remove "escape_string" from imports

We use prepared statements, so no need to have this.

BonfaceKilz committed:
gn3: db: sort imports

BonfaceKilz committed:
tests: test_phenotype: Rename to test_db

BonfaceKilz committed:
tests: test_db: Rename TestPhenotype

BonfaceKilz committed:
gn3: db: Use correct DATACLASSMAP entry from metadata_audit

BonfaceKilz committed:
gn3: db: Make "WHERE" clause optional

* gn3/db/__init__.py (fetchone): Make "WHERE" an Optional arg.

BonfaceKilz committed:
gn3: metadata_audit: Make props for MetadataAudit class optional

BonfaceKilz committed:
gn3: db: Add "fetchall" method.

BonfaceKilz committed:
tests: test_db: Add a test-case for fetchall

BonfaceKilz committed:
gn3: db: Add "id_" property to metadata_audit class and mapping

BonfaceKilz committed:
gn3: db: Fix how columns from tables is resolved

BonfaceKilz committed:
Rename json_data column to json_diff_data

BonfaceKilz committed:
db: test_db: Add extra variable in test data for "id"

Alexanderlacuna committed:
initial commit for queries perfomance tests

Alexanderlacuna committed:
add types and dataset names

Alexanderlacuna committed:
add tests to setup packages

Alexanderlacuna committed:
pep8 formatting

Alexanderlacuna committed:
rename perf query file

Alexanderlacuna committed:
refactor perf query functions

Alexanderlacuna committed:
remove ununsed variables

Alexanderlacuna committed:
minor fixes

zsloan committed:
Fixed spelling of coeffient to coefficient

Alexanderlacuna commented:
@zsloan  Looks Good. Do you capture the results by writing to a CSV file or the stdout?

zsloan commented:
> @zsloan Looks Good. Do you capture the results by writing to a CSV file or the stdout?

The R script writes to a CSV file and GN3 reads in the contents of that file (two files if permutations are used). The API call from GN2 receives the results as JSON.

zsloan committed:
Resolve mypy errors in computations/rqtl.py

zsloan committed:
add ignore_missing_imports for numpy in mypy.ini since it complains about that and we seem to have done the same thing for scipy

zsloan committed:
Fixed pylint issues

zsloan committed:
Change test_compose_rqtl_command in test_commands.py to have different arguments to hopefully avoid github's pylint 'duplicate code' complaint

Changed some parameters in test_compose_rqtl_cmd to avoid pylint complaining about duplicate code

zsloan committed:
Fixed test_compose_rqtl_command test; forgot to include Rscript in the command

zsloan committed:
Fixed test_rqtl.py to include Rscript in the command

zsloan committed:
Fixed file type from json to csv for test_generate_rqtl_command

zsloan committed:
Fixed another error where test_generate_rqtl_command didn't include the filename argument (not sure why running unit tests locally doesn't detect this)
